pkg/discovery/discovery.go

```diff
@@ -226,3 +235,102 @@ func (d *tidbDiscovery) DiscoverDM(advertisePeerUrl string) (string, error) {
 	delete(currentCluster.peers, podName)
 	return fmt.Sprintf("--join=%s", strings.Join(mastersArr, ",")), nil
 }
+
+func (d *tidbDiscovery) VerifyPDEndpoint(pdURL string) (string, error) {
+	// save a copy of pdURL for failure
+	copyPDEndpoint := pdURL
+	pdEndpoint := parsePDURL(pdURL)
+
+	ns := os.Getenv("MY_POD_NAMESPACE")
+	tc, err := d.cli.PingcapV1alpha1().TidbClusters(ns).Get(pdEndpoint.tcName, metav1.GetOptions{})
+	if err != nil {
+		klog.Errorf("Failed to get the tidbcluster when verifying PD endpoint")
+		return pdURL, err
+	}
+
+	noScheme := false
+	if len(pdEndpoint.scheme) == 0 {
+		noScheme = true
+		if tc.IsTLSClusterEnabled() {
+			pdEndpoint.scheme = "https"
+		} else {
+			pdEndpoint.scheme = "http"
+		}
+		pdURL = fmt.Sprintf("%s://%s", pdEndpoint.scheme, pdURL)
+	}
+
+	if d.pdEndpointHealthCheck(tc, pdURL, pdEndpoint.pdMemberName) {
+		// the input PD endpoint works normally
+		klog.Infof("The PD endpoint: %s is healthy", pdURL)
+		if noScheme {
+			return fmt.Sprintf("%s:%s", pdEndpoint.pdMemberName, pdEndpoint.pdMemberPort), nil
+		}
+		return pdURL, nil
+	}
+
+	var returnPDMembers []string
+	var returnPDMember string
+	returnPDMembers = append(returnPDMembers, copyPDEndpoint)
+	for _, pdMember := range tc.Status.PD.PeerMembers {
+		if pdMember.Health {
+			if noScheme {
+				returnPDMember = fmt.Sprintf("%s:%s", pdMember.Name, pdEndpoint.pdMemberPort)
```

Comment:
I think we should get the URL from `pdMember.ClientURL` because the `pdMember.Name` and `pdEndpoint.pdMemberPort` may not be reliable.